### PR TITLE
feat: Consider business hours while generating the reports

### LINKED ISSUE #3446: feat: Consider business hours while generating the reports
**Is your enhancement request related to a problem? Please describe.**
We are not considering the week's off days now while generating the report which affects the average resolution time.

**Describe the solution you'd like**
We should consider the week's off days that we mention in the settings while generating the report.

**Describe alternatives you've considered**
While generating the report deduct the number of off days.

**Additional context**
Refer: https://app.chatwoot.com/app/accounts/1/conversations/11059

### Todo:
- [x] #4057
- [ ] Add a value_in_business_hours, event_start_time, event_end_time. In the migration, use nil as existing value.
- [ ] Compute value_in_business_hours while creating the event. If all_day, then value = value_in_business_hours. If not all_day, compute the difference in hours
- [ ] Add toggle near the graph to see values_in_business_hours.

### PR DESCRIPTION
**Pull request by aswindevps**
## Description

This change allows the user to filter reports based on the business hours configured in the inbox. Below are the metrics supported by the filter.

1. avg_first_response_time
2. avg_resolution_time

Fixes #3446

## Type of change

- [ ] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

https://user-images.githubusercontent.com/5848740/161263438-95910402-17a7-4c3b-9c08-f69d128a84a4.mp4

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -787,6 +801,9 @@
   add_foreign_key "campaigns", "inboxes", on_delete: :cascade
   add_foreign_key "contact_inboxes", "contacts", on_delete: :cascade
   add_foreign_key "contact_inboxes", "inboxes", on_delete: :cascade
+  add_foreign_key "conversation_participants", "accounts", on_delete: :cascade
+  add_foreign_key "conversation_participants", "conversations", on_delete: :cascade
+  add_foreign_key "conversation_participants", "users", on_delete: :cascade
```
Comment (pranavrajs) [inline db/schema.rb:806]: 👃

File: db/schema.rb [hunk 2]
```diff
@@ -337,6 +337,17 @@
...
+  create_table "conversation_participants", force: :cascade do |t|
+    t.bigint "account_id", null: false
+    t.bigint "user_id", null: false
+    t.bigint "conversation_id", null: false
+    t.datetime "created_at", precision: 6, null: false
+    t.datetime "updated_at", precision: 6, null: false
+    t.index ["account_id"], name: "index_conversation_participants_on_account_id"
+    t.index ["conversation_id"], name: "index_conversation_participants_on_conversation_id"
+    t.index ["user_id"], name: "index_conversation_participants_on_user_id"
+  end
```
Comment (pranavrajs) [inline db/schema.rb:349]: 👃

Comment (aswindevps) [reply db/schema.rb]: Removed

File: db/schema.rb (hunk 1 above)
Comment (aswindevps) [reply db/schema.rb]: Removed

**Review by muhsin-k — DISMISSED** [review]
Front end code looks good to me

File: app/models/reporting_event.rb [hunk 3]
```diff
@@ -2,15 +2,18 @@
...
+#  id                      :bigint           not null, primary key
+#  event_end_time          :datetime
+#  event_start_time        :datetime
+#  name                    :string
+#  value                   :float
+#  value_in_business_hours :float
+#  created_at              :datetime         not null
+#  updated_at              :datetime         not null
+#  account_id              :integer
+#  conversation_id         :integer
```
Comment (tejaswinichile) [inline app/models/reporting_event.rb:14]: Are considering the old records how would it affect those?  Or the value can be default null?

Comment (aswindevps) [reply app/models/reporting_event.rb:14]: Yes. For the old records, the below fields value would be null. 

1. value_in_business_hours
2. event_start_time
3. event_end_time

@sojan-official

**Review by tejaswinichile — APPROVED** [review]
LGTM 👍🏼

**Review by muhsin-k — APPROVED** [review]

**Pull request merged**